### PR TITLE
Update block to be a class - Closes #6581

### LINKED ISSUE #6581: Update block to be a class
### Description

Update block and block header to be a class with below interface.
This is required to support `getAsset` and `setAsset`

```
class Block {
  public static fromBytes(value: Buffer): Block;
  public getBytes(): Buffer;
  public get header() BlockHeader;
  public get payload() []Transaction;
}

class BlockHeader {
  public static fromBytes(value: Buffer): BlockHeader;
  public getBytes(): Buffer;
  public toJSON(): BlockHeaderJSON;
  public validate(): void;
  public getSigningBytes(): Buffer;
  public sign(key: Buffer);
  public getAsset(moduleID: number): Buffer;
  public setAsset(moduleID: number, value: Buffer): void;
  public get generatorAddress(): Buffer;
  public get id(): Buffer;
  public get version(): number;
  public get previousBlockID(): Buffer;
  public get timestamp(): Buffer;
  public get stateRoot(): Buffer;
  public get transactionRoot(): Buffer;
  public get signature(): Buffer;
}
```

### Acceptance Criteria
- Add unit test for the added classes

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?

This PR resolves #6581

### How was it solved?

- Created two classes `Block` and `BlockHeader` 

### How was it tested?

- Added unit tests

### CODE REVIEW

File: elements/lisk-chain/src/block_header.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	public sign(key: Buffer, networkIdentifier: Buffer) {
+		return signDataWithPrivateKey(TAG_BLOCK_HEADER, networkIdentifier, this.getSigningBytes(), key);
+	}
+
+	public getAsset(moduleID: number): Buffer | undefined {
+		return this._assets.find(a => a.moduleID === moduleID)?.data;
+	}
+	public setAsset(moduleID: number, value: Buffer): void {
+		const asset = this.getAsset(moduleID);
+		if (asset) {
```
Comment (shuse2) [inline elements/lisk-chain/src/block_header.ts:126]: should we overwrite?

File: elements/lisk-chain/src/block_header.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		this.stateRoot = stateRoot;
+		this.transactionRoot = transactionRoot;
+		this._assets = [...assets];
+
+		this._signature = signature;
+		this._id = id;
+	}
+
+	public static fromBytes(value: Buffer): BlockHeader {
+		return new BlockHeader(codec.decode<BlockHeaderAttrs>(blockHeaderSchema, value));
```
Comment (shuse2) [inline elements/lisk-chain/src/block_header.ts:81]: probably `id` should be calculated here as well before assigning

File: elements/lisk-chain/src/block_header.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	public toJSON(): BlockHeaderJSON {
+		return codec.toJSON(blockHeaderSchema, this._getBlockHeaderProps());
+	}
+
+	public validate(): void {
+		const headerWithoutAsset = {
+			...this._getBlockHeaderProps(),
+			asset: Buffer.alloc(0),
```
Comment (shuse2) [inline elements/lisk-chain/src/block_header.ts:95]: we don't have asset =)

File: elements/lisk-chain/src/block_header.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		};
+		// Validate block header
+		const errors = validator.validate(blockHeaderSchema, headerWithoutAsset);
```
Comment (shuse2) [inline elements/lisk-chain/src/block_header.ts:98]: no need to specifically change it to without asset. we can have the schema `data` as bytes

File: elements/lisk-chain/src/block_header.ts (hunk 1 above)
Comment (nazarhussain) [reply elements/lisk-chain/src/block_header.ts:126]: Approach 1: Different modules can set asset on different logic, so we can overwrite asset. 
Approach 2: Let user handle logic on their side and set the asset once for a module. 

I believe Approach 2 is safe.

Comment (shuse2) [reply elements/lisk-chain/src/block_header.ts:189]: I was thinking a case where a module want to set in `initBlock` and update in `sealBlock`, but i guess it's not necessary for now

File: elements/lisk-chain/src/block_header.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		}
+	}
+
+	public getSigningBytes(): Buffer {
+		const blockHeaderBytes = codec.encode(signingBlockHeaderSchema, this._getSigningProps());
+
+		return blockHeaderBytes;
+	}
+
+	public sign(key: Buffer, networkIdentifier: Buffer) {
```
Comment (shuse2) [inline elements/lisk-chain/src/block_header.ts:164]:
```suggestion
	public sign(networkIdentifier: Buffer, privateKey: Buffer) {
```

**Review by shuse2 — CHANGES_REQUESTED** [review]
Overall LGTM, just minor things

**Review by shuse2 — APPROVED** [review]

File: elements/lisk-chain/src/block.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { BlockHeader } from './block_header';
+import { blockSchema } from './schema';
+import { Transaction } from './transaction';
+
+interface BlockAttrs {
+	header: Buffer;
+	payload: Buffer[];
+}
+
+export class Block {
```
Comment (ishantiw) [inline elements/lisk-chain/src/block.ts:25]: The issue says it should have 2 more methods,

```js
public get header() BlockHeader;
public get payload() []Transaction;
```

File: elements/lisk-chain/test/unit/block_header.spec.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright © 2019 Lisk Foundation
```
Comment (ishantiw) [inline elements/lisk-chain/test/unit/block_header.spec.ts:2]: if its a new file then update license
```suggestion
 * Copyright © 2021 Lisk Foundation
```

File: elements/lisk-chain/test/unit/block_header.spec.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	'previousBlockID',
+	'stateRoot',
+	'transactionRoot',
+	'generatorAddress',
+	'assets',
+];
+
+describe('block_header', () => {
+	describe('signingBlockHeaderSchema', () => {
+		it('should be valid schema without signature or id', () => {
```
Comment (ishantiw) [inline elements/lisk-chain/test/unit/block_header.spec.ts:60]:
```suggestion
		it('should be valid schema without signature and id', () => {
```

File: elements/lisk-chain/src/block_header.ts [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { objects } from '@liskhq/lisk-utils';
+import { signDataWithPrivateKey, hash } from '@liskhq/lisk-cryptography';
+import { codec } from '@liskhq/lisk-codec';
+import { validator, LiskValidationError } from '@liskhq/lisk-validator';
+import { TAG_BLOCK_HEADER } from './constants';
+
+export const signingBlockHeaderSchema = {
```
Comment (ishantiw) [inline elements/lisk-chain/src/block_header.ts:21]: should we move these schemas to `schema.ts`?

File: elements/lisk-chain/src/block_header.ts [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+	readonly previousBlockID: Buffer;
+	readonly timestamp?: number;
+	readonly stateRoot?: Buffer;
+	readonly transactionRoot?: Buffer;
+	readonly assets: ReadonlyArray<BlockHeaderAsset>;
+	signature?: Buffer;
+	id?: Buffer;
+}
+
+export interface BlockHeaderJSON {
```
Comment (ishantiw) [inline elements/lisk-chain/src/block_header.ts:95]: if we are exporting then maybe move it to `types.ts`

File: elements/lisk-chain/src/block.ts (hunk 6 above)
Comment (nazarhussain) [reply elements/lisk-chain/src/block.ts:25]: Not required to to have getters as the `header` and `payload` both are public readonly properties.

File: elements/lisk-chain/src/block_header.ts (hunk 9 above)
Comment (nazarhussain) [reply elements/lisk-chain/src/block_header.ts:21]: Yes we should, when we integrate new block header. If done now, will break a lot of things in the framework and other places.

File: elements/lisk-chain/src/block_header.ts (hunk 10 above)
Comment (nazarhussain) [reply elements/lisk-chain/src/block_header.ts:95]: Same as above.

**Review by ishantiw — APPROVED** [review]

**Pull request merged**